artifactory/commands/transferfiles/state/transfersnapshot.go

```diff
@@ -0,0 +1,101 @@
+package state
+
+import (
+	"github.com/jfrog/jfrog-cli-core/v2/utils/reposnapshot"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"sync"
+	"time"
+)
+
+var saveRepoSnapshotMutex sync.Mutex
+
+type SnapshotActionFunc func(rts *RepoTransferSnapshot) error
+
+var snapshotSaveIntervalMin = snapshotSaveIntervalMinDefault
+
+const snapshotSaveIntervalMinDefault = 10
+
+// RepoTransferSnapshot handles saving and loading the repository's transfer snapshot.
+type RepoTransferSnapshot struct {
+	snapshotManager   reposnapshot.RepoSnapshotManager
+	lastSaveTimestamp time.Time
+	// This boolean marks that this snapshot continues a previous run. It allows skipping certain checks if it was not loaded, because some data is known to be new.
+	loadedFromSnapshot bool
+}
+
+// Runs the provided action on the snapshot manager, and periodically saves the repo state and snapshot to the snapshot dir.
+func (ts *TransferStateManager) snapshotAction(action SnapshotActionFunc) error {
+	if ts.repoTransferSnapshot == nil {
+		return errorutils.CheckErrorf("invalid call to snapshot manager before it was initialized")
+	}
+	if err := action(ts.repoTransferSnapshot); err != nil {
+		return err
+	}
+
+	now := time.Now()
+	if now.Sub(ts.repoTransferSnapshot.lastSaveTimestamp).Minutes() < float64(snapshotSaveIntervalMin) {
+		return nil
+	}
+
+	if !saveRepoSnapshotMutex.TryLock() {
+		return nil
+	}
+	defer saveRepoSnapshotMutex.Unlock()
+
+	ts.repoTransferSnapshot.lastSaveTimestamp = now
+	if err := ts.repoTransferSnapshot.snapshotManager.PersistRepoSnapshot(); err != nil {
+		return err
+	}
+
+	return saveStateToSnapshot(&ts.TransferState)
```

Comment:
It will not block the transfer process, only the single thread who called the action that triggered the saving. Other threads will go on without interruptions.

Reply:
Sounds good!